File: benchmarks/cross/cases/storage/whole_read.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

func main() {
	if len(os.Args) != 3 {
		fmt.Fprintln(os.Stderr, "usage: whole_read REPETITIONS FIXTURE")
		os.Exit(2)
	}
	repetitions, err := strconv.Atoi(os.Args[1])
	if err != nil || repetitions <= 0 {
		fmt.Fprintln(os.Stderr, "repetitions must be positive")
		os.Exit(2)
	}
	var totalBytes uint64
	var checksum uint64
	for i := 0; i < repetitions; i++ {
		data, err := os.ReadFile(os.Args[2])
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			os.Exit(1)
		}
		totalBytes += uint64(len(data))
		checksum += uint64(len(data))
		checksum += uint64(data[0]) + uint64(data[len(data)-1])
	}
	if totalBytes%uint64(repetitions) != 0 {
		fmt.Fprintln(os.Stderr, "inconsistent byte count")
		os.Exit(1)
	}
	fmt.Printf("bytes=%d reps=%d checksum=%d\n",
		totalBytes/uint64(repetitions), repetitions, checksum)
}

File: benchmarks/cross/cases/storage/whole_read.cpp
#include <cstdint>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

std::vector<unsigned char> read_all(const std::string& path) {
  std::ifstream source(path, std::ios::binary | std::ios::ate);
  if (!source) {
    throw std::runtime_error("cannot open fixture: " + path);
  }
  const auto end = source.tellg();
  if (end < 0) {
    throw std::runtime_error("cannot determine fixture size");
  }
  std::vector<unsigned char> data(static_cast<std::size_t>(end));
  source.seekg(0);
  if (!source.read(reinterpret_cast<char*>(data.data()), end)) {
    throw std::runtime_error("failed while reading fixture");
  }
  return data;
}

int main(int argc, char** argv) {
  if (argc != 3) {
    std::cerr << "usage: whole_read REPETITIONS FIXTURE\n";
    return 2;
  }
  try {
    const int repetitions = std::stoi(argv[1]);
    if (repetitions <= 0) {
      throw std::runtime_error("repetitions must be positive");
    }
    std::uint64_t total_bytes = 0;
    std::uint64_t checksum = 0;
    for (int i = 0; i < repetitions; ++i) {
      const auto data = read_all(argv[2]);
      total_bytes += data.size();
      checksum += data.size();
      checksum += data.front() + data.back();
    }
    if (total_bytes % static_cast<std::uint64_t>(repetitions) != 0) {
      throw std::runtime_error("inconsistent byte count");
    }
    std::cout << "bytes=" << total_bytes / repetitions
              << " reps=" << repetitions
              << " checksum=" << checksum << '\n';
  } catch (const std::exception& error) {
    std::cerr << error.what() << '\n';
    return 1;
  }
}
